Tenants, External Users, and tenant based JWT with Audit Events

edpaget opened the pull request:
### Description

Most customers doing self-service embedded analytics with Metabase have a multi-tenant setup, where multiple tenants (customer organizations) are co-located on a shared Metabase instance, and rely on data isolation (row/column security, DB routing or connection impersonation) to ensure each user can only see their relevant data.

To achieve this, customers can use the same building blocks we have for the internal analytics use case:  collections, users, groups, and collection and data permissions.

To push the self-service embedded analytics project to production, surfacing the embeds (ex. full app embedding iframe) is just the tip of the iceberg. Figuring out how to use the raw building blocks to implement tenancy is complex, and onboarding and offboarding tenants, and their users requires complex API integrations to provision/deprovision and configure collections, groups, group permissions, and group memberships programmatically. This can easily turn into a multi-week project.

Tenants aims to make all of this much simpler for 80% of our embedding customers.

### What's in

- Availability: The feature is available to Pro and Enterprise customers via the `tenants` feature flag in the token
- Enabling/Disabling the feature:
    - Enable the feature by changing MB from single-tenant to multi-tenant mode in Admin/Settings/People.
    - Disable the feature in the same place. Note that disabling the feature will cause all tenant users to be mass disabled.
    - When the feature is enabled the UI changes a bit
        - the tenants related features appear
        - the “All users” group for internal appears as “All internal users”
- Migration path:
    - A user can be transitioned from internal to external or viceversa using the user endpoint in the API. This is important because internal users and tenant users are stored physically in the same table and we don't allow hard deleting users or creating another user with the same email address (regardless if of a different type)
- Tenants:
    - A tenant is an entity that maps to an external organization. In a typical SaaS application these would be the customers.
    - A tenant has a display name and a slug, which will be used for JWT SSO (see below). The slug is also propagated as an attribute, so it can be used for sandboxing as well.
    - A tenant can optionally have tenant attributes too. Tenant users inherit attributes from their tenant and can also have user attributes.
    - A tenant user must always be a member of a tenant.
- Tenant groups:
    - Tenant groups are a special type of group for tenant users. members (tenant users) can be long to one or multiple different tenants.
    - There is an “All tenant users” group to facilitate certain actions, like granting all tenant users via permission for a specific shared collection or setting up sandboxing for all tenant users based on tenant attributes or slug
- Shared collections:
    - These are collections that are curated by internal users (who can curate these can be controlled via regular permissions) and shared with tenant users (can also be controlled via permissions for tenant groups, but can be view or blocked only, and never curate)
    - There can be zero or many shared collections.
    - Shared collections can be shared with all tenant users (via all tenant users group). This is a one-time action that will make it easy to ensure any new tenant user automatically has access to it.
    - Shared collections can also be shared to a subset of tenant users via tenant groups.
- Tenant collections
    - When a tenant is provisioned, a tenant collection is provisioned as well under the hood. Every member of that tenant can read and write (curate permissions) to that collection. We may add support for more fine grained permissions support in the future.
- Personal tenant collections
    - Every tenant user gets their own personal collection which they can read and write to
- Using tenants w/ JWT SSO
    - we are adding a new JWT claim `@tenant` . MB will do a look up of tenants by slug and sign in or provision a new tenant user and make the user a member of that tenant or provision a new tenant if it's missing, provided the feature is enabled
    - if the feature is disabled, we will reject to sign new tenant users
    - we also reserve JWT claims prefixed with @ moving forward for Metabase and specific features. Customers should not use these for other purposes as they will collide. We didn't see any cloud customers using any @-prefixed attributes in Hosting Insights.
    - It's not yet possible to add tenant attributes via JWT SSO, so to avoid additional REST API calls, sandboxing can be configured to use the slug value. That way just by using SSO, you can create new users in new tenants and have sandboxing work right away without having to hit the API.
    - Group syncing can be used as well for tenant groups. It needs to be enabled under JWT SSO. Group syncing labels are not needed. Just pass the exact string of the tenant name, if it's a match and group syncing is enabled, MB will make the user a member of that group.
- Remote sync
    - The remote sync feature can be used to sync shared collections across MB instances
    - This can be done by enabling syncing for specific shared collections, individually.
    - Tenant collections and personal tenant collections cannot be synced
- Guardrails
    - tenant users can never access internal content, and can only ever access
        - their own personal collection
        - the tenant collection from the tenant they are members of
        - any shared collections they have access via tenant groups (only view or blocked)
    - a tenant user is always a member of one single tenant
    - a tenant user cannot be a member of a regular group
    - In JWT SSO
      - if the `@tenant` claim is present, but doesn't match the tenant the user belongs to, or the user with that email is an internal user, an error is raised
      - if the `@tenant` claim is missing, but the user with that email is a tenant user, an error is raised

### How to verify

1. Enable the feature
    1. Head to admin/people, click on the cog in the top right, change user strategy to Multi tenant
2. Create a shared collection
    1. Click on "Tenants” in the sidebar. Click on "Create shared collection” in the top right. Enter a name, ex: "Shared collection"
    2. Note: we want to automatically create a first tenant collection, but this hasn't been implemented yet
3. Configure permissions
    1. Open admin settings / permissions
    2. Tenant collections - By default, all tenant users will have view access to the shared collections - no need to touch this now but understand you can change here which internal groups can curate or view and which tenant groups can view shared collections
    3. Data - Ensure users are sandboxed
        1. Click on data tab, select all tenant users, sample DB
        2. Select orders table, row/column security. Map column user ID to `@tenant.slug`. Create queries "Query builder only”, downloads "10K rows".
        3. Set all other tables to Blocked.
        4. Save changes
4. Create your first tenant
    1. Go to admin / people, click on tenants in the sidebar, “New tenant”  top right
    2. Enter a name for the tenant. Use `"1”` as tenant slug or some other value that you know will be present the DB, in the column you used for sandboxing
5. Create a tenant user
    1. Open tenant users in the sidebar
    2. “Create tenant user” in the top right. Copy/paste the 
6. To “sample” the experience as an end-user, you have different options
    1. Open an incognito window in your browser and sign in as the tenant user you just created  
    2. Or use any of the embedding samples (interactive, EAJS, SDK), but make sure you add `@tenant`  claim to the SSO JWT, with the value you used to create a tenant (in reality, it will also provision a new tenant if it doesn't find a tenant with that slug 
    3. …
7. Remote sync
    1. You can use remote sync to move the shared collections across instances.
    2. In admin settings, enable remote sync
    3. Once enabled, in remote sync, you can toggle remote sync for each shared collection individually
    
### Demo

_Upload a demo video or before/after screenshots if sensible or remove the section_

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

johnswanson committed:
Tenants, External Users, and tenant based JWT with Audit Events

Allow the creation of tenant users

While working on this, I realized that my approach to the "All External
Users" group and renaming the "All Users" group was untenable.

I had been planning to just update these things in the database (create
AEU, rename AU to AIU) in the setter function for the `:use-tenants`
setting. The issue here is caching: we don't want to need to actually
reach out to the database every time we need to fetch the All Users or
All External Users, but it would also be quite unfortunate if one
Metabase instance in a cluster had an outdated view of, e.g. the ID of
"All External Users".

(Although - I'm not really sure how much we use the *name* of "All
Users" when we fetch `(all-users)`, so that might be worth a look.)

For now, I've modified things to just update things dynamically in the
`GET /api/permissions/group` and `GET /api/permissions/group/:id`
endpoints: we show/hide the "All External Users" group, and rename the
"All Users" group, based on whether Tenants are enabled.

Allow the creation of tenants

Listing tenants / tenant members will be a followup PR

lint

doh missing require

fix tests

add a test for enabling tenants changing groups

fix tests, slight change to tenant creation API

kick ci

fix another test

give minimum permissions to ext groups

fix more tests

impl user stategy setting ui

more API endpoints

add more tests, fix a very silly migration

slug validation

fix mariadb/mysql

lint migrations file

kondo

fix test

replace tenant members API with `/user?tenant_id=`

also realized that the `/user/recipients` endpoint needed to be modified
for tenants: tenant users should not see recipients that are not part of
the same tenant. (reminder for later: we will need to add safeguards
around actually creating those pulses, not just assume the API won't
send bad recipients!!)

IF you are an admin, `/api/user` respects a `tenant_id` query param and returns:

- "normal" (non-tenant) users if it's `nil`

- specific-tenant users if not

If you are *not* an admin, `/api/user` ignores the `tenant_id` passed
and only returns users in the same tenant.

`/api/user/recipients` will:

- return ALL users if you are a superuser (both tenant and non-tenant)

- return only users in the SAME tenant as you if not

doh

tenant crud wip

integrate endpoints, fix type errors

Allow getting all external users

previously you could do only

```
GET /api/user                => internal users
GET /api/user?tenant_id=null => internal users
GET /api/user?tenant_id=123  => users in tenant 123
```

now you can do

```
GET /api/user                  => internal users
GET /api/user?tenant_id=null   => internal users
GET /api/user?tenant_id=123    => users in tenant 123
GET /api/user?tenancy=all      => internal + external users
GET /api/user?tenancy=internal => internal users
GET /api/user?tenancy=external => external users (in all tenants)
```

If you specify both `tenancy` and `tenant_id` (regardless of the values
passed) you'll get a 400.

fix my bad rebase

fix bad rebase more

remove #p

mark tenants as `is_active` or not

initial external users management

remove permissions check on the trash endpoint

I don't know why this was added in the first place (I mean, I think I
added it, but I don't know why I did).

The trash is always accessible to everyone. We gate the visibility of
things *in* the trash, but the trash itself is always readable.

no personal collection for tenant users

remove tenant query inside oss

tenant user routes

make it so that you cannot remove users from the external users group

do not send email to tenant users

no login for inactive tenants

disable collection permissions, don't show sent email modal on user success with SMTP

remove all users group has higher permissions warnings for external users group

moving is external checks behind plugins

unit test cleanup, add simple test

read only application permissions for external users

set tenant on login with JWT

incredibly hacky tenant attribute

separate `:attributes` from `:login_attributes`

thinking here is that `login_attributes` are the things you set on the
user, where `attributes` are all the attributes of the user including
things like tenant slug or maybe someday locale etc.

fix build

fix token check

create a tenant collection in a migration

this works, but the tenant collection is always visible to admins, even
if the setting is off.

hide tenant collection from normal users

fixing build and FE rebase

Delete stray "howdy"

Lol, sorry about that.

fix tests

throw instead of silent failure on no tenant

superuser can't be tenant

fix up tenant collections

tenant collection now includes subcollections.

Tenant collections are hidden if the feature isn't enabled.

tenant collections marked in API

bugfix and current user can write any coll

don't add `attributes` key to all users

just the current user on login

invalidate groups when changing user strategy

read only user

{{ tenant_slug }} user attribute

fix be test

rip out tenant collection stuff on BE

fix BE tests

lint

fix magic string

no tenantifying existing users for now

also clean up a complicated, confusing fn.

starting e2e

fixing e2e, adjusting @tenant user attribute

fix permissions editor, add tenant.name e2e test

allow more than one user per tenant

:facepalm:

allow getting only active tenants

*actually* don't replace existing values with nil

getting e2e ready for tenant user provisioning

tenants management ux

e2e coverage on sandboxing, deactivating tenants

replace {{tenant_slug}} with `@tenant.slug`

filter out other-tenant users from pulse creation

rename any existing `@.+` user attributes

prohibit new user attributes starting with `@`

fix migration

ugh, forgot to actually commit this file

add more tests

kondo

use the `tenants` feature

fix a test - missed one `:advanced-permissions`

e2e test coverage, updating tenant attribute to @tenant.slug

Partial Type Fixes

types green

unit tests

adding groups to e2e test, maybe fixing tenant feature check?

Add tenants to tracked snowplow features

Fixing Admin ui

Modernizing tenants page

separate external group in permissions sidebar

permissions editor updates

hardcode that *everyone* has trash access

shame... bad imports

adjusting Main Navbar. using globe icon

CR response

fix tests

fix attributes when tenants are disabled

lock up the 'all external users' collection

tenants route guard

add tests for missing feature and oss

fixing some unit tests, and type errors

set token features in default.cy.snap

unit and e2e test adjustment

another pass on e2e

fix e2e and build

fix a bad merge

Green CI?

reset token after restoring DB

move a test to fix the test

kondo

Move `/api/ee/tenants` to `/api/ee/tenant`

PR Feedback, part 1

move route functions into ee

ADM-1105 permissions editor adjustments for external users

ADM 1104 do not show invite button on external users when there is no tenant

adjusting unit tests

fix no results message for deactivated users page

center no results message

Fix migrations IDs / related tests

adjusting no results message for tehants

fixing unit tests, adding e2e test for permissions

chore: add timestamp columns to tenants (#60575)

* chore: add timestamp columns to tenants

* fixup! add hooks

* kick ci

Auto provision tenants, respect tenants feature (#60571)

deactivate tenant users along with their tenants (#60552)

* deactivate tenant users along with their tenants

Tenant users from deactivated tenants were already prevented from
logging in (we checked the activation status of their tenant) but for
bookkeeping it's probably best to actually deactivate the users
themselves along with their tenants, and reactivate them if the tenant
is reactivated.

* invalidate users when updating a tenant, adjust external users list screen

* e2e test

---------

Co-authored-by: Nick Fitzpatrick <nickfitz.582@gmail.com>

audit events for tenants (#60647)

johnswanson committed:
Revert "Delete unused string util from metabase lib (#64389)"

This reverts commit 0af38ef8f021726985eacccddffca8f0b43233f8.

johnswanson committed:
fix the build

johnswanson committed:
fix tests

heypoom committed:
fix rebase error in invite button

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix another test

johnswanson committed:
Fix another couple tests

johnswanson committed:
and more

johnswanson committed:
and more

heypoom committed:
Close the modal to create shared tenant collection after it is created (#66049)

johnswanson committed:
Fix more tests, nearly all e2e tests passing

johnswanson committed:
a couple more tests

johnswanson committed:
couple more tests

johnswanson committed:
last (???) few tests

heypoom committed:
Remove edit permissions in ellipsis menu when viewing shared tenant collection (#66042)

heypoom committed:
Suggest external users only when adding tenant group members (#66046)

johnswanson committed:
Dedicated Tenant Collections

edpaget committed:
feat: separate namespace for tenant collections (#66102)

edpaget committed:
fixup! FE build

johnswanson committed:
More fixes

johnswanson committed:
Tests and more fixes

edpaget committed:
feat: add remote-sync support for tenant collections (#66272)

* feat: add remote-sync support for tenant collections

* Update permissions.clj

* fixup! tests

* fixup! enterprise tests

* fixup! mysql/h2

edpaget committed:
EMB-1011: add default read perms for namespace

johnswanson committed:
Fix rebase error

johnswanson committed:
Fix tests (#66403)

* Fix tests

Not exactly sure how these got merged, I think maybe they were hidden by
GH issues.

* Fix kondo

* fix apparent rebase error

edpaget committed:
fixup! EMB-1065 fix namespace filtering

edpaget committed:
fixup! some bad perms tests

edpaget committed:
fixup! remote sync dep checking across collections

edpaget committed:
fixup! remove blank test

edpaget committed:
fixup! missing tenant setting

edpaget committed:
fixup! add premium feature declaration

edpaget committed:
fixup! get test working with h2

edpaget committed:
fixup! revert h2 fix

edpaget committed:
fixup! revert h2 fix

edpaget committed:
fixup! real h2 fix

johnswanson committed:
rename

johnswanson committed:
Fix query for shared tenant collections

johnswanson committed:
Give tenant users a personal collection

edpaget committed:
fixup! UXW-2386 don't pass `tables` model into tenant root

edpaget committed:
fixup! UXW-2436 build correct path for namespaced collections

edpaget committed:
Remove test of non-existent fn

heypoom committed:
Fix failing frontend unit and e2e tests in tenants feature branch (#66243)

Co-authored-by: John Swanson <john.d.swanson@gmail.com>

heypoom committed:
fix failing frontend ESLint errors

heypoom committed:
Prefix users and groups as "Internal" in admin settings when tenants are active (#66169)

heypoom committed:
Update design of the user strategy modal for tenants (#66182)

Co-authored-by: Mahatthana (Kelvin) Nomsawadi <me@bboykelvin.dev>

heypoom committed:
Add multi-tenancy setup guide to embedding hub (#66113)

heypoom committed:
Update design of the tenants listing and internal user page (#66193)

johnswanson committed:
Fix BE tests

johnswanson committed:
another be test

johnswanson committed:
kondo/fmt

heypoom committed:
Add specific error messages for JWT tenant claim errors (#66172)

johnswanson committed:
fix running jwt tests standalone

edpaget committed:
feat: per collection setting (#66475)

* feat: per collection setting

* feat: remove unsynced collections

* fixup! tests

* fixup! lints

* fixup! test

edpaget committed:
Remove duplicated tenant permissions

johnswanson committed:
handle disabling feature more cleanly

when tenants is turned off:

- disable all tenant users

- reject any logins from tenant users

- reject any JWTs that specify a tenant

edpaget committed:
Add tenant entity picker e2e tests. Fix one bug surfaced

heypoom committed:
uxw-2456: remove drop shadows from dedicated tenant collection cards

heypoom committed:
Hide tenant collections permissions tab when tenants are disabled (#66537)

heypoom committed:
Hide create shared tenant collection button on sidebar for non-admin users (#66538)

heypoom committed:
Hide tenant collections sidebar section when no collections exist for non-admins (#66539)

heypoom committed:
Fix dedicated tenant collection root in collection picker as tenant user (#66549)

edpaget committed:
EMB-2459: Fix restrict where collections can be moved to

heypoom committed:
fix frontend type-check: only set namespace if it is defined

johnswanson committed:
EMB-1093: Fix not passing a tenant attribute

Fixes [EMB-1093: If you enable multi-tenants and don't add JWT
attribute, it will
break](https://linear.app/metabase/issue/EMB-1093/if-you-enable-multi-tenants-and-dont-add-jwt-attribute-it-will-break)

edpaget committed:
Fixup working e2e entity picker tests

edpaget committed:
fixup! mysql namespace tests

edpaget committed:
fixup! tenant group mapping tests

edpaget committed:
fixup! ws

johnswanson committed:
tree endpoint allows specifying multiple nses

johnswanson committed:
fix broken test

heypoom committed:
Show dedicated tenant collections in collection pickers for admins (#66583)

heypoom committed:
feat(sdk-bundle): add tenants ee plugin (#66586)

edpaget committed:
fixup! bad ns clause audit feature is off

edpaget committed:
fixup! UXW-2473 support non-string attributes

heypoom committed:
Hide tenant collections section when there are no visible shared collections (#66597)

edpaget committed:
Add tenant cards to search

edpaget committed:
fixup! tenant collection cleanup

npretto committed:
EMB-1082 rename strings in the ui (#66603)

npretto committed:
make tenant field required (#66516)

johnswanson committed:
change `tenant` key to `@tenant`

- if the feature is off, you can still log in if you have the `@tenant`
key set. It will be silently ignored, but a warning will be logged.

- if the feature is on, `@tenant` is a drop-in replacement for `tenant`.

- we'd previously prevented setting user attributes that start with `@`.
This updates `remove-invalid-attributes` to do the same with JWT/SAML
attributes.

johnswanson committed:
add SAML tests

johnswanson committed:
Expose SAML attribute for tenant

johnswanson committed:
External users can't use MB if feature disabled

Plus add tests for:

- tenant user, feature disabled

- tenant user, feature unavailable

johnswanson committed:
Add missing test files

heypoom committed:
Multi-stage embedding data picker should query models and questions from tenant collections (#66598)

heypoom committed:
Fix external group tooltip referencing All Internal Users (#66654)

johnswanson committed:
Allow conversion of tenant <-> internal users

Upon conversion, all previous group memberships are destroyed and the
user is added to either the "All External Users" or "All Users" groups
accordingly.

edpaget committed:
feat: implementing new tenant/remote-sync designs (#66567)

johnswanson committed:
fix missing dep in test file

and a tiny removal of an unnecessary `dissoc`

heypoom committed:
fixup!: broken embedding page due to bad rebase conflict resolution

heypoom committed:
fixup!: fix type errors

heypoom committed:
fixup!: fix frontend jest unit test errors

edpaget committed:
fixup! Move search tests requiring EE to EE code

edpaget committed:
Fix copy test

npretto committed:
add confirmation warning when disabling tenants (#66697)

johnswanson committed:
fix test

npretto committed:
more renaming

heypoom committed:
fixup!: failing tenants.cy.spec.js test and rename "All tenant users" migration

edpaget committed:
fixup! whitespace fixes

edpaget committed:
Merge branch 'master' into tenants-base

edpaget committed:
Remove serialization for dropped column

edpaget committed:
fixup! missing migration

edpaget committed:
fix other wrongly ordered migration

edpaget committed:
fixup! Ordering more migrations

edpaget committed:
fixup! other tests

johnswanson committed:
fix test

johnswanson committed:
fix another test

johnswanson committed:
api response schemas for tenant endpoints

johnswanson committed:
kondo

johnswanson committed:
one more kondo

edpaget committed:
fixup! bad test query

edpaget committed:
fixup! popover to use the right way of passing args

edpaget committed:
Fix push changes modal

edpaget committed:
fixup! correct token for entity picker tests

edpaget committed:
Fix tenants e2e test with correct attribute

edpaget committed:
fixup! tenants tests

edpaget committed:
fixup! remove test for old behavior

npretto committed:
fix type on events, removed unused string union values

npretto committed:
update MainNavSharedCollections.unit.spec.tsx to use the new name

npretto committed:
fix linting issue unused import

npretto committed:
remove test file not on master anymore

johnswanson committed:
fix permissions select

johnswanson committed:
fix dashboard filters autowiring e2e test

edpaget committed:
fix formating of module.edn

edpaget committed:
remove useless test file

iethree commented on e2e/test/scenarios/embedding/modular-embedding-settings.cy.spec.ts:
```diff
@@ -0,0 +1,32 @@
+
+  it("should link to tenants page when tenants are enabled", () => {
+    cy.visit("/admin/embedding/modular");
+    H.updateSetting("use-tenants", true);
```
again, should probably update this setting before the page load

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/source.clj:
```diff
@@ -59,25 +59,32 @@
     (throw entity))
-  (u/prog1 {:path (remote-sync-path opts entity)
-            :content (yaml/generate-string entity {:dumper-options {:flow-style :block :split-lines false}})}
+  (u/prog1 (if (string? entity)
```
this feels bad

iethree commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/GitSyncControls/GitSyncControls.tsx:
```diff
@@ -0,0 +1,237 @@
+
+  return (
+    <>
+      <div className={styles.pill} data-testid="git-sync-controls">
```
this is pretty special, but maybe Mantine's `<Button.Group />` would work instead of something so bespoke?

https://mantine.dev/core/button/#buttongroup

iethree commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/SharedTenantCollectionsList/SharedTenantCollectionsList.tsx:
```diff
@@ -0,0 +1,137 @@
+            checked={isChecked}
+            onChange={(e) => onToggle(collection, e.currentTarget.checked)}
+            disabled={!canWrite || isReadOnly}
+            aria-label={t`Sync ${collection.name}`}
```
this string could use context, it's gonna be hard to translate `sync {0}` without context in a lot of language

```
c("{0} is the name of a metabase collection").t(`Sync ${collection.name}`)

nvoxland commented on enterprise/backend/test/metabase_enterprise/tenants/core_test.clj:
```diff
@@ -0,0 +1,348 @@
+      (mt/with-temporary-setting-values [use-tenants true]
+        (is (empty? (tenants/login-attributes {:tenant_id 99999})))))))
+
+#_(deftest attribute-structure-ee-with-user-and-tenant-attributes-test
```
Should all these commented tests be re-enabled or removed?

iethree committed:
Merge branch 'master' into tenants-base

# Conflicts:
#	frontend/src/metabase-types/api/collection.ts
#	resources/migrations/058_update_migrations.yaml

iethree committed:
add enum for namespaces

edpaget commented on enterprise/backend/test/metabase_enterprise/remote_sync/settings_test.clj:
this test should be removed

edpaget commented on enterprise/backend/test/metabase_enterprise/tenants/api_test.clj:
```diff
@@ -0,0 +1,638 @@
+                (is (= (str "/" tenant-collection-id "/" child-a-id "/")
+                       (t2/select-one-fn :location :model/Collection :id child-b-id)))))))))))
+
+(deftest cannot-move-regular-collection-into-tenant-namespace-test
```
Because these aren't testing the tenants api could they got in another test ns as a sibling to this one (like collections_test.clj)

edpaget commented on enterprise/backend/test/metabase_enterprise/tenants/permissions_test.clj:
```diff
@@ -0,0 +1,98 @@
+                                                                      :is_tenant_group true}
+                       :model/PermissionsGroup {normal-group-id :id} {:name "Normal Group"
+                                                                      :is_tenant_group false}]
+
```
blank lines

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/components/MainNavSharedCollections/MainNavSharedCollections.tsx:
```diff
@@ -0,0 +1,164 @@
+  const [modalOpen, setModalOpen] = useState(false);
+  const isTenantsEnabled = useSelector((state) =>
+    getSetting(state, "use-tenants"),
+  );
```
there a nicer hook for this `useSetting("use-tenants")`

iethree commented on enterprise/frontend/src/metabase-enterprise/api/tenants.ts:
```diff
@@ -0,0 +1,78 @@
+        method: "GET",
+        url: `/api/collection/root/items`,
+        params: {
+          models: ["collection"],
```
I think only collections can be in the root for tenants?

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/components/TenantCollectionPermissionsPage/TenantCollectionPermissionsPage.tsx:
```diff
@@ -0,0 +1,152 @@
+  route: Route;
+};
+
+function TenantCollectionPermissionsPageView({
```
do we really need a whole new page for this? couldn't we leverage the existing page and pass it some props? or is the logic really that different?

edpaget replied:
I can make it use the existing page with props, but I wasn't sure where standard line is for adding props vs combining the same components in different ways as a new component is in the FE.

iethree replied:
yeah that's a familiar story, permissions are pretty redux-spaghetti-heavy. thanks for looking at it.

taking a holistic approach to untangling permissions a bit is probably the better course

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/components/TenantForm/TenantForm.tsx:
```diff
@@ -0,0 +1,107 @@
+  submitText?: string;
+}
+
+function slugify(str: string): string {
```
why not `import { slugify } from "metabase/lib/formatting";` ?

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/components/TenantSpecificCollectionsItemList/TenantSpecificCollectionsItemList.tsx:
```diff
@@ -0,0 +1,62 @@
+  );
+};
+
+const getTenantCollections = (
```
can't we just query collections in the `tenant-specific` namespace from the api instead of constructing a fake data structure from the list of tenants?

edpaget committed:
fixup! remove toggle

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/containers/NewTenantModal/NewTenantModal.tsx:
```diff
@@ -0,0 +1,40 @@
+      name: vals.name ?? "",
+      slug: vals.slug ?? "",
+    }).unwrap();
+    dispatch(addUndo({ message: t`Tenant creation successful` }));
```
the `useToast` hook is much nicer

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/containers/TenantActivationModal.tsx:
```diff
@@ -0,0 +1,62 @@
+        onClose={onClose}
+        onConfirm={async () => {
+          await updateTenant({ id: tenantId, is_active: false });
+          dispatch(addUndo({ message: t`Tenant deactivated` }));
```
consider using `useToast` instead of dispatch/addUndo

edpaget commented on enterprise/frontend/src/metabase-enterprise/remote_sync/analytics.ts:
```diff
@@ -3,7 +3,7 @@ import { trackSimpleEvent } from "metabase/lib/analytics";
   triggeredFrom,
 }: {
-  triggeredFrom: "sidebar" | "admin-settings";
+  triggeredFrom: "sidebar" | "admin-settings" | "app-bar";
```
```suggestion
  triggeredFrom: "admin-settings" | "app-bar";
```

edpaget commented on frontend/src/metabase-types/analytics/event.ts:
```diff
@@ -473,7 +473,7 @@ export type ClickActionPerformedEvent = ValidateEvent<{
 export type RemoteSyncBranchSwitchedEvent = ValidateEvent<{
   event: "remote_sync_branch_switched";
-  triggered_from: "sidebar" | "admin-settings";
+  triggered_from: "sidebar" | "admin-settings" | "app-bar";
```
```suggestion
  triggered_from: "admin-settings" | "app-bar";
```

iethree commented on enterprise/frontend/src/metabase-enterprise/tenants/utils/colors.ts:
```diff
@@ -0,0 +1,14 @@
+import { color } from "metabase/lib/colors";
+import type { Tenant } from "metabase-types/api";
+
+const groupColorPalette = [
```
these should be CSS vars and not use the deprecated `color` function

iethree commented on frontend/src/metabase/admin/components/RelatedSettingsSection/constants.ts:
```diff
@@ -40,8 +47,22 @@ export const getModularEmbeddingRelatedSettingItems =
       name: t`Appearance`,
       to: "/admin/settings/appearance",
     },
+    ...(isTenantsFeatureAvailable
```
this code should probably be in the tenants plugin

edpaget committed:
Update EntityPickerModal.tsx

iethree commented on frontend/src/metabase/admin/people/components/GroupDetail.tsx:
```diff
@@ -149,6 +155,19 @@ export const GroupDetail = ({
+        <p>
+          {t`All tenant users belong to the ${getGroupNameLocalized(
+            group,
+          )} group and can't be removed from it. Setting permissions for this group is a great way to
```
this is tenant-feature-specific behavior and should be in the tenants plugin

iethree commented on frontend/src/metabase/admin/people/components/PeopleNav.tsx:
```diff
@@ -7,35 +7,60 @@ import {
+          label={isUsingTenants ? t`Internal Groups` : t`Groups`}
           icon="group"
         />
+        {isUsingTenants && (
```
this should probably be in the tenants plugin

iethree reviewed (commented):
This is just the first chunk of my front-end review, but I've spent 5 hours on this and can't look at it any more today. Hopefully I can get through the rest tomorrow

Overall observations:
- This is looking really good and I haven't found any major problems, just some cleanup to do
- I'm really pleasantly surprised at how well tested much of this 👏

edpaget commented on resources/migrations/056_update_migrations.yaml:
```diff
@@ -212,6 +212,7 @@ databaseChangeLog:
                   constraints:
                     nullable: true
 
+
```
```suggestion
```

edpaget commented on src/metabase/permissions/models/permissions.clj:
```diff
@@ -483,16 +489,17 @@
   (check-is-modifiable-collection collection-or-id)
   (when (perms-group/is-tenant-group? group-or-id)
-    (throw (ex-info (tru "Tenant Groups cannot have write access to any collections.") {})))
+    (throw (ex-info (tru "Tenant groups cannot have write access to any collections.") {})))
```
Should these say external groups? I've lost track on whether we always call these 'External groups' or Tenant groups.

edpaget commented on src/metabase/premium_features/token_check.clj:
```diff
@@ -73,7 +73,11 @@
              (let [result (binding [t2.conn/*current-connectable* nil]
-                            (t2/count :model/User :is_active true :type :personal))]
+
+                            ;; Because we need this count *during* token checks, this uses `t2/table-name` to avoid
```
Should we fix this problem?

edpaget commented on test/metabase/users/models/user_test.clj:
```diff
@@ -534,4 +535,73 @@
                 "user_attr" "user_value"}
-               (:attributes result)))))))
+               (:attributes result))))))
+  (testing "tenant attributes"
```
This test should be split into multiple deftests

edpaget committed:
Merge branch 'master' into tenants-base

edpaget committed:
Remove commented tests

edpaget committed:
Update e2e test based on feedback

edpaget committed:
Split add-attributes-test

edpaget committed:
Update test/metabase/collections_rest/api_test.clj

edpaget committed:
Fix copy cmd model list

edpaget committed:
Remove :put handling code for is_remote_synced

edpaget committed:
Update src/metabase/premium_features/settings.clj

edpaget committed:
Remove extraneous regex in api def

edpaget committed:
Update resources/migrations/056_update_migrations.yaml

edpaget committed:
Update frontend/src/metabase-types/analytics/event.ts

edpaget committed:
UXW-2536 popup remote-sync modal after request

edpaget committed:
Remove the "sidebar" event type

edpaget committed:
Improve some backend tests

edpaget committed:
Merge branch 'master' into tenants-base

npretto committed:
refactor: remove useAdminSetting and use settings already in the
component

edpaget committed:
Fix remote-sync e2e tests

npretto committed:
same for the jwt form

npretto committed:
move error message to Messages

npretto committed:
variant filled for apply button

johnswanson committed:
convert attributes to strings on read/write

In practice there is no difference between the value `true` and the
value `"true"`, so let's keep things simple by making all attributes
maps of string->string.

Fixes [UXW-2550: Cannot edit tenant users with non-string attributes](https://linear.app/metabase/issue/UXW-2550/cannot-edit-tenant-users-with-non-string-attributes)

edpaget committed:
fixup extra dialog on permspagelayout

npretto committed:
fw={500} => medium

npretto committed:
add c() to t`Sync {collection.name}`

npretto committed:
useSetting("use-tenants") in MainNavSharedCollections

johnswanson committed:
fix broken tests

johnswanson committed:
fix the other broken test

edpaget committed:
Remove tenant collection app permission

edpaget committed:
Fix embedding sdk collection type

edpaget committed:
Convert tenants e2e test to typescript

npretto committed:
move setupRemoteSyncEndpoints to server-mocks folder

npretto committed:
actually remove it from test-utils

edpaget committed:
Small BE fixes

npretto committed:
use usePrevious from react-use

edpaget committed:
Remove custom root tenant collection hook

edpaget committed:
Change group detail to use text

npretto committed:
fix sdk type

edpaget committed:
Redid git controls with button.group

edpaget committed:
fixup! bad test

iethree commented on frontend/src/metabase/admin/people/containers/GroupsListingApp.tsx:
```diff
@@ -11,26 +13,43 @@ import {
-    await createGroup(group).unwrap();
+    await createGroup({
+      ...group,
+      is_tenant_group: external, // TODO, make the API accept this
```
did this get done? if so we should remove this comment

iethree commented on frontend/src/metabase/admin/permissions/components/PermissionsSelect/PermissionsSelect.jsx:
```diff
@@ -59,57 +59,53 @@ export const PermissionsSelect = memo(function PermissionsSelect({
-      triggerElement={selectedOptionValue}
-    >
-      {({ onClose }) => (
+    <Popover opened={opened} onChange={setOpened} disabled={isDisabled}>
```
👏 nice popover improvements

iethree commented on frontend/src/metabase/admin/utils.js:
```diff
@@ -41,3 +42,18 @@ export const RedirectToAllowedSettings = connect(
+    allowRedirectBack: false,
+    authenticatedSelector: (state) =>
+      getAdminPaths(state)?.find((path) => path.key === "people") != null &&
+      MetabaseSettings.get("use-tenants"),
```
😭😭😭😭 please don't use MetabaseSettings unless you have to use the window object before we have a redux store loaded. I think you should be able to use `getSetting` selector here

iethree commented on frontend/src/metabase/common/components/Pickers/CollectionPicker/components/CollectionPicker/tests/common.unit.spec.tsx:
```diff
@@ -0,0 +1,86 @@
+    ).toHaveAttribute("data-active", "true");
+  });
+
+  it("should allow selecting, but not navigating into collections without children", async () => {
```
it looks like all of these were copied from `CollectionPicker.unit.spec.tsx` except the last test `should allow disabling certain items in the collection picker`

after copying that test, `CollectionPicker.unit.spec.tsx` should be deleted

iethree commented on frontend/src/metabase/common/components/Pickers/CollectionPicker/hooks.ts:
```diff
@@ -26,6 +32,16 @@ const personalCollectionsRoot: CollectionPickerItem = {
+  location: "/",
+  description: "",
+  here: ["collection"],
+  below: ["collection"],
```
this might cause bugs with the questionPicker, it might not let you navigate into this folder since it thinks there's no data there

iethree commented on frontend/src/metabase/common/components/Pickers/MiniPicker/components/MiniPicker/test/ee.unit.spec.tsx:
```diff
@@ -35,7 +35,7 @@ const LIBRARY_METRICS_COLLECTION = createMockCollectionItem({
 describe("library", () => {
   it("should default to the library if it's available", async () => {
-    fetchMock.get("/api/ee/library", LIBRARY_COLLECTION);
+    fetchMock.get("path:/api/ee/library", LIBRARY_COLLECTION);
```
we shouldn't be setting up bare mocks like this, there should be a helper function in the api mocks folder for exactly this reason: we can define the path once.

edpaget replied:
I think this was one introduced by the library work and this was fixing it for flakes here.

iethree reviewed (approved):
Frontend code is acceptable. Didn't see anything that *must* block, but there's a lot of things in the nature of cleanup that we really should do (FIXME/TODO comments, etc.)

I tried to indicate the high priority items with 😭 emojis

edpaget committed:
fixup! tenant logic and components into plugins

edpaget committed:
Update frontend/src/metabase/lib/string.js

Co-authored-by: Ryan Laurie <30528226+iethree@users.noreply.github.com>

edpaget commented:
@iethree I can't reply [this comment](https://github.com/metabase/metabase/pull/66661#discussion_r2612061600) for some reason, but the reason I wrote it like that was that we needed a new wrapper to show collection dirty state now that remote-synced collections just displayed in the regular collection section of the sidebar. When i tried to put the oss version without the dirty indicators in the oss plugin implementation running tests caused a double import of the clojurescript standard library which meant they wouldn't run.

edpaget committed:
Merge branch 'master' into tenants-base

edpaget committed:
fixup! bad merge

edpaget committed:
Rearrange last unit test for CollectionPicker

edpaget committed:
fixup! PR feedback

edpaget committed:
fixup! handle nils and seqs in attributes

edpaget committed:
Merge branch 'master' into tenants-base

edpaget committed:
fixup! kondos

edpaget committed:
fixup! better stringify

edpaget committed:
Add missing maybe to login attributes schema

edpaget committed:
fixup! yaml tests

edpaget committed:
fixup! sequential check

edpaget committed:
fixup! Stringify ldap attributes

edpaget committed:
Fix ldap tests
